Customizability

patrickfatrick opened the pull request:
Added Font Awesome and Devicons to the project, as well as functionality in main.js to support customizing the icons in the Brackets preferences file using the following format:

"file-icons.icons": [
	{
			"extension": "html",
			"icon": "fa fa-code",
                        "color": "#E84D49",
			"size": 16
        }
]

The only required attribute is "extension", everything else will default to what's already in the extension if not specified. Next thing I'd like to do is add the ability to add your own extensions if not already supported, but for now you can customize any extension in main.js (including aliases). Oh, and I also added HAML.

UPDATE: Now includes the ability to create your own extension. If you list an extension in the preferences file that's not included in the fileInfo object, an entry will be added. See second screenshot.

![brackets-customizable-icons](https://cloud.githubusercontent.com/assets/5255524/6654224/e8d97506-ca7a-11e4-921e-bd73d41de8c5.png)

![brackets-customizable-icons-2](https://cloud.githubusercontent.com/assets/5255524/6656271/1c169dc6-caea-11e4-9141-36db3be90cc3.png)

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
 
+	// Load preferences
+	var PreferencesManager = brackets.getModule("preferences/PreferencesManager"),
+		prefs = PreferencesManager.getExtensionPrefs("file-icons");
```
Use two separate var statements here

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
+	prefs.definePreference("icons", "array", []);
+
+	// Declare variables
+	var iconArray;
```
Call this variable `userIcons`, place it directly under `var fileInfo = {}` and rename `fileInfo` to `defaultIcons`

ivogabe replied:
And make it an object, just like fileIcons/defaultIcons

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
+
+	// Declare variables
+	var iconArray;
+	var iconObject = {};
```
Don't use a global variable for this

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
+	}
+
 	function addIcon(extension, icon, color, size) {
+		checkPrefs(extension, icon, color, size);
```
Don't call checkPrefs here, I think there should be one array with the default icons and one with the user icons. This way we can reload the user icons array if the user changes his settings, without needing to restart Brackets.

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
+
 	function addAlias(extension, other) {
 		fileInfo[extension] = fileInfo[other];
+		checkPrefs(extension, fileInfo[extension].icon, fileInfo[extension].color, fileInfo[extension].size);
```
Same as line 37

ivogabe commented on main.js:
```diff
@@ -37,96 +78,99 @@ define(function (require, exports, module) {
+	addIcon('eot', 'ion-social-tumblr', '#b36908');
+	addIcon('woff', 'ion-social-tumblr', '#7f4bb2');
+	addIcon('woff2', 'ion-social-tumblr', '#7f4bb2');
+	addIcon('otf', 'ion-social-tumblr', '#7f4bb2');
```
Use a different color for this icon. Woff & woff2 can share the same color

ivogabe commented on main.js:
```diff
@@ -190,6 +234,27 @@ define(function (require, exports, module) {
 	addIcon('d', 'ion-contrast', '#960000');
 	addIcon('r', 'ion-ios-analytics', '#8495C0');
 
+	function extensionSearch(needle, haystack) {
```
Remove this (line 237-257). Instead write a function that looks for a certain extension, by first checking whether it exists in `userIcons`, and if not, check it in `defaultIcons`

ivogabe replied:
You can add that to the `provider` function instead of a new one actually.

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
+	// Declare variables
+	var iconArray;
+	var iconObject = {};
+	var iconArray = prefs.get("icons");
```
Create a new function that loads the settings, and call it here and when the settings are changed (https://github.com/adobe/brackets/wiki/Preferences-System#change-events)

ivogabe commented:
I'd prefer the following format:
```json
"brackets-icons.icons": {
  "html": {
    "icon": "fa fa-code",
    "color": "#E84D49",
    "size": 16
  }
}
```
- `brackets-icons` instead of `file-icons`, since this extension is called Brackets-Icons.
- Map instead of array: less configuration, less code to read settings.

ivogabe commented on main.js:
```diff
@@ -1,16 +1,57 @@
 	var fileInfo = {};
 
+	// Load preferences
+	var PreferencesManager = brackets.getModule("preferences/PreferencesManager"),
```
Use single quotes (`'`)

ivogabe commented:
Thanks for the contribution! I've added some comments to the code. Can you add a new contribution.md file, which describes how to use this feature? You may add a reference to that file from the readme.

patrickfatrick commented:
Thanks for your comments. Very helpful as I know it's not super pretty :P. I'll work on this stuff soon (maybe not today though).

patrickfatrick commented:
Somethings changed:

-Changed preferences to use object instead of array.

-Highly condensed and cleaned up the amount of code. Now everything is contained in
function applyPrefs. Basically we have an object called “icons” and one
called “userIcons”. addIcon writes to the icons object, and then
afterward the function applyPrefs updates or adds to the icons object.

-Auto-reloads preferences if they’re changed. One quirk I wasn’t able
to figure out is that I can’t get the file tree or working set to
auto-refresh. If you click on any item inside they will refresh though.

-Added a separate md file to describe usage.

Let me know what you think!

ivogabe commented on main.js:
```diff
@@ -190,6 +207,26 @@ define(function (require, exports, module) {
 	addIcon('r', 'ion-ios-analytics', '#8495C0');
 
+	//Look up preferences and apply changes
+	function applyUserPrefs() {
```
Don't fill the `icons` object with user icons. With this approach, reloading settings will fail when the user removed an extension from the settings file. You should check in the provider function (line 247) whether the icon exists in the `userIcons` object, and otherwise check it in the `icons` object.

ivogabe commented:
Looks good!

patrickfatrick commented:
That's a good point.

I've got everything in the provider function now and seems to be working well. 

I might need some guidance on the best way to get it to reload though. So it does reload but you do have to click on something in the working set and file tree to get the icons to change over. (And it does now also work with removing an entry in the preferences file, from my testing). Wrapped all of the addIcon/addAlias calls in function called loadDefaults which I'm also calling in prefs.on('change') in the case of an item, that also has a default, being changed or removed.  I've got some commented out code in the prefs.on('change') handler where I tried some things out but it didn't work out.

ivogabe commented:
I don't think the loadDefaults function is necessary. Can you do the following:
- Revert commit "Changes to reloading"
- Replace `var icon = defaultIcons[ext]; [...] var data = icon ? icon : getDefaultIcon(ext);` with:
```javascript
var data;

if (userIcons.hasOwnProperty(ext)) {
  data = userIcons[ext];
} else if (defaultIcons.hasOwnProperty(ext)) {
  data = defaultIcons[ext];
} else {
  data = getDefaultIcon(ext);
}
```
Keeping the default icons size or color is really needed, it will probably give some unexpected behavior. I'll take a look at refreshing the tree when it's merged.

patrickfatrick commented:
I've removed the loadDefaults function, reverted back the prefs.on("change") method to where it was (let me know if you want to just get rid of that functionality entirely until we're able to get it working better, as that would take a couple seconds to take out :).

Changed the provider function to what you're looking for, so now an entire icon spec (extension, icon, color, size) will need to be specified in preferences; you can't change one or two values in an existing default icon. Made changes to the customizability readme file to reflect this change.

ivogabe commented on main.js:
```diff
@@ -1,16 +1,28 @@
 			color: color,
 			size: size
-		};
+		}
```
Add semicolon again

ivogabe commented on readme.md:
```diff
@@ -1,17 +1,18 @@
 Brackets Icons
 ==============
-This extension adds file icons (from the [Ionicons](https://github.com/driftyco/ionicons) project) to the brackets file tree. 
+This extension adds file icons (from the [Ionicons](http://ionicons.com), [Font Awesome](https://fortawesome.github.io/Font-Awesome/), and [Devicons](https://vorillaz.github.io/devicons/#/main) projects) to the brackets file tree. 
```
Revert this, you can reference the other 2 projects in customizability.md.

ivogabe commented:
Thanks!
